gitlab/auth.go

```diff
@@ -0,0 +1,276 @@
+/*
+Copyright 2020 The Flux CD contributors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package gitlab
+
+import (
+	"fmt"
+	"net/http"
+
+	"github.com/fluxcd/go-git-providers/gitprovider"
+	gogitlab "github.com/xanzy/go-gitlab"
+	"golang.org/x/oauth2"
+)
+
+const (
+	// DefaultDomain specifies the default domain used as the backend.
+	DefaultDomain = "gitlab.com"
+)
+
+// ClientOption is the interface to implement for passing options to NewClient.
+// The clientOptions struct is private to force usage of the With... functions.
+type ClientOption interface {
+	// ApplyToGitlabClientOptions applies set fields of this object into target.
+	ApplyToGitlabClientOptions(target *clientOptions) error
+}
+
+// clientOptions is the struct that tracks data about what options have been set.
+type clientOptions struct {
+	// clientOptions shares all the common options
+	gitprovider.CommonClientOptions
+
+	// AuthTransport is a ChainableRoundTripperFunc adding authentication credentials to the transport chain.
+	AuthTransport gitprovider.ChainableRoundTripperFunc
```

Comment:
Github has `EnableConditionalRequests` etc we don't have that in Gitlab?

Reply:
Found this gitlab issue https://gitlab.com/gitlab-org/gitlab-foss/-/issues/61559 that tracks some of the progress and this page in the development docs https://docs.gitlab.com/ee/development/polling.html.
The same mechanism by using the `gregjones/httpcache` transport worked in the tests, but still needs to be verified in the integration to WKP. Adding the same ConditionalRequests option to test :+1:

Reply:
Cool! We can also leave it out and make a follow up issue 👍

Reply:
Just to track it somwhere.

Reply:
It's added :+1: test pending